File: EmptyProject/WorkflowsTrainingTwo/WorkflowsTraining/Parsing/Parser.cs
using WorkflowsTraining.Lexing;
using WorkflowsTraining.Parsing.Statements;
using WorkflowsTraining.Parsing.Statements.Interfaces;

namespace WorkflowsTraining.Parsing;

public class Parser(StreamReader sourceCodeStream)
{
    private readonly TokenStream _stream = Lexer.Tokenize(sourceCodeStream);

    public ProgramNode Parse()
    {
        IList<IStatementNode> statements = [];
        while (!_stream.EndOfFile())
        {
            statements.Add(StatementParser.ParseStatement(_stream));
        }

        return new ProgramNode(statements);
    }
}

File: EmptyProject/WorkflowsTrainingTwo/WorkflowsTrainingCPlus/Parser.cpp
#include "TokenStream.cpp"
#include "Lexer.cpp"
#include "ProgramNode.cpp"

namespace WorkflowsTraining {
    namespace Parsing {
        class Parser {
        private:
            WorkflowsTraining::Lexing::TokenStream _stream;

        public:
            Parser(std::ifstream& sourceCodeStream)
                : _stream(WorkflowsTraining::Lexing::Lexer::Tokenize(sourceCodeStream)) {
            }

            ProgramNode Parse() {
                std::vector<std::unique_ptr<Statements::IStatementNode>> statements;

                while (!_stream.EndOfFile()) {
                    statements.push_back(Statements::StatementParser::ParseStatement(_stream));
                }

                return ProgramNode(std::move(statements));
            }
        };
    }
}